src/main/java/com/wanderersoftherift/wotr/abilities/attachment/AbilityTracker.java

```diff
@@ -0,0 +1,123 @@
+package com.wanderersoftherift.wotr.abilities.attachment;
+
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.Multimap;
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.codecs.RecordCodecBuilder;
+import com.wanderersoftherift.wotr.abilities.Ability;
+import com.wanderersoftherift.wotr.abilities.TrackedAbilityTrigger;
+import com.wanderersoftherift.wotr.init.WotrAttachments;
+import com.wanderersoftherift.wotr.init.WotrRegistries;
+import com.wanderersoftherift.wotr.item.ability.AbilityModifier;
+import com.wanderersoftherift.wotr.modifier.WotrEquipmentSlot;
+import com.wanderersoftherift.wotr.serialization.AttachmentSerializerFromDataCodec;
+import net.minecraft.core.Holder;
+import net.minecraft.nbt.Tag;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.LivingEntity;
+import net.minecraft.world.level.Level;
+import net.neoforged.neoforge.attachment.IAttachmentHolder;
+import net.neoforged.neoforge.attachment.IAttachmentSerializer;
+import org.jetbrains.annotations.Nullable;
+
+public class AbilityTracker {
+
+    public static final IAttachmentSerializer<Tag, AbilityTracker> SERIALIZER = new AttachmentSerializerFromDataCodec<>(
+            Data.CODEC, AbilityTracker::new, AbilityTracker::data);
+
+    private final Multimap<Holder<TrackedAbilityTrigger.Type<?>>, TrackedAbility> abilities = ArrayListMultimap
+            .create();
+    private final IAttachmentHolder holder;
+
+    public AbilityTracker(IAttachmentHolder holder) {
+        this(holder, null);
+    }
+
+    public AbilityTracker(IAttachmentHolder holder, @Nullable Data data) {
+        this.holder = holder;
+    }
+
+    public static AbilityTracker forEntity(Entity entity) {
+        return entity.getData(WotrAttachments.ABILITY_TRACKER.get());
+    }
+
+    public static AbilityTracker forEntityNullable(Entity entity) {
```

Comment:
There is a `getExistingData()` method on IAttachmentHolder that achieves the same effect.